ext/opcache/shared_alloc_mmap.c

```diff
@@ -40,6 +40,105 @@
 # define MAP_HUGETLB MAP_ALIGNED_SUPER
 #endif
 
+#if defined(__x86_64__) && defined(__linux__)
+/* On Linux OS, this function returns the start address of the first PHP segment
+   through parsing /proc/self/maps file; on any other OSes, simply returns NULL.
+*/
+static void* get_php_start_addr(void)
+{
+	/* Used to parse each field for the line in /proc/self/maps file */
+	long unsigned int start, end, offset, inode;
+	char perm[5], dev[10], name[MAXPATHLEN];
+	int ret;
+	FILE *f = NULL;
+
+	f = fopen("/proc/self/maps", "r");
+	if (!f) {
+		return NULL;
+	}
+
+	/* Only get the start address of the first PHP segment */
+	ret = fscanf(f, "%lx-%lx %4s %lx %9s %ld %s\n",
+					&start, &end, perm, &offset, dev, &inode, name);
+	fclose(f); f=NULL;
+
+	/* Expect to get seven fields */
+	if (7 == ret) {
+		return (void*)start;
+	}
+	return NULL;
+}
+
+/*
+	This function tries to create segment starting from address prior to
+	the first segment of PHP process and returns the newly-allocated address,
+	or MAP_FAILED for failure.
+
+	If huge page is supported and requested_size meets conditions, we will first
+	try to get memory from huge pages; if failed, fall back to try 4KB pages.
+*/
+static void* create_preferred_segments(size_t requested_size)
+{
+	int flags = PROT_READ | PROT_WRITE, fd = -1;
+	void *p;
+
+	void *php_start_addr = NULL; /* start address of first PHP segment */
+	void *preferred_mmap_addr = NULL; /* for mmap() */
+
+	long unsigned int huge_page_size = 2 * 1024 * 1024; /* 2MB page size */
+	long unsigned int ordinary_page_size = 4 * 1024; /* 4KB page size */
+
+	php_start_addr = get_php_start_addr();
+
+	/* Do mmap allocation with preferred address only if
+	   1) PHP starts from above 4GB address
+	   2) Enough free space before PHP segments
+	      (Reserve one page hole for memory address alignment)
+	*/
+
+#if defined(MAP_HUGETLB)
+	/* Try to get memory from huge pages */
+	if (php_start_addr &&
+		((long unsigned int) php_start_addr > UINT_MAX) &&
```

Comment:
We could use `UINT32_MAX` and `uintptr_t` here

Reply:
Yes, good idea and will fix this.